services/src/util/extractors.rs

```diff
@@ -0,0 +1,165 @@
+use futures::future::Ready;
+use std::fmt;
+
+use actix_http::Payload;
+use actix_web::dev::JsonBody;
+use actix_web::{FromRequest, HttpRequest};
+use futures::future::LocalBoxFuture;
+use futures::FutureExt;
+use serde::de::DeserializeOwned;
+use validator::Validate;
+
+use crate::handlers::ErrorResponse;
+
+use super::server::handle_json_payload_error;
+
+/// A Json extractor that validates the content after deserialization
+#[derive(Debug)]
+pub struct ValidatedJson<T>(pub T);
+
+impl<T> ValidatedJson<T> {
+    /// Deconstruct to an inner value
+    pub fn into_inner(self) -> T {
+        self.0
+    }
+}
+
+impl<T> AsRef<T> for ValidatedJson<T> {
+    fn as_ref(&self) -> &T {
+        &self.0
+    }
+}
+
+impl<T> std::ops::Deref for ValidatedJson<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        &self.0
+    }
+}
+
+impl<T> FromRequest for ValidatedJson<T>
+where
+    T: DeserializeOwned + Validate + 'static,
+{
+    type Error = actix_web::Error;
+    type Future = LocalBoxFuture<'static, Result<Self, Self::Error>>;
+
+    #[inline]
+    fn from_request(req: &HttpRequest, payload: &mut Payload) -> Self::Future {
+        JsonBody::new(req, payload, None, true)
+            .limit(2_097_152) // Default limit is 2MB in actix-web
+            .map(|res: Result<T, _>| match res {
+                Ok(data) => {
+                    data.validate()
+                        .map(|_| ValidatedJson(data))
+                        .map_err(|validation_errors| {
+                            ErrorResponse {
+                                error: "ValidationError".to_string(),
+                                message: validation_errors_to_string(&validation_errors),
+                            }
+                            .into()
+                        })
+                }
+                Err(e) => Err(handle_json_payload_error(e)),
+            })
+            .boxed_local()
+    }
+}
```

Comment:
What is the idea behind not wrapping `web::Json`?

You could say 

```rust
pub struct ValidatedJson<T>(web::Json<T>);
```

Then, you could just forward `FromRequest` and then call validated, right?

Reply:
the idea of forwarding to web::Json FromRequest is nice. However, the method returns a future, but we cannot await this future here, because we are not in an async function.

Furthermore, the `pub struct ValidatedJson<T>(pub T);` just wraps the `T` value for returning it to the users and it is constructed in the method. There is no benefit in wrapping a web::Json here, because the user wants the `T` value and the parsing and validation happens beforehand.

Reply:
You also return a `Future`, so you can just await, map, etc. I do not see this point.

The nice thing would just be to get the `actix` implementation and all updates automatically.

The compiler should optimize the nested newtypes.